File: apps/flow-backend/src/routers/workflow.ts
/**
 * Copyright (c) 2025 Bytedance Ltd. and/or its affiliates
 * SPDX-License-Identifier: MIT
 */

/**
 * Workflow CRUD router (protected — requires a Bearer API key).
 *
 * - Secrets in `document` are encrypted on write, decrypted on read.
 * - `update` uses optimistic concurrency: clients send the `version` they
 *   loaded; stale writes are rejected with CONFLICT.
 * - `list` omits the (potentially large) document body; fetch `get` for the doc.
 */
import { z } from 'zod';
import { TRPCError } from '@trpc/server';

import { protectedProcedure, router } from '../trpc.js';
import { encryptDocumentSecrets, decryptDocumentSecrets } from '../lib/secrets.js';
import { prisma } from '../db.js';

const documentSchema = z.record(z.any());

export const workflowRouter = router({
  list: protectedProcedure
    .input(z.object({ search: z.string().optional() }).optional())
    .query(({ ctx, input }) =>
      prisma.workflow.findMany({
        where: {
          ownerId: ctx.user.id,
          ...(input?.search ? { name: { contains: input.search } } : {}),
        },
        select: {
          id: true,
          name: true,
          version: true,
          createdAt: true,
          updatedAt: true,
        },
        orderBy: { updatedAt: 'desc' },
      })
    ),

  get: protectedProcedure.input(z.object({ id: z.string() })).query(async ({ ctx, input }) => {
    const wf = await prisma.workflow.findFirst({
      where: { id: input.id, ownerId: ctx.user.id },
    });
    if (!wf) throw new TRPCError({ code: 'NOT_FOUND' });
    return {
      ...wf,
      document: decryptDocumentSecrets(wf.document as never),
    };
  }),

  create: protectedProcedure
    .input(z.object({ name: z.string().min(1), document: documentSchema }))
    .mutation(async ({ ctx, input }) => {
      const wf = await prisma.workflow.create({
        data: {
          name: input.name,
          ownerId: ctx.user.id,
          document: encryptDocumentSecrets(input.document) as never,
        },
      });
      return { id: wf.id, version: wf.version };
    }),

  update: protectedProcedure
    .input(
      z.object({
        id: z.string(),
        name: z.string().min(1).optional(),
        document: documentSchema.optional(),
        version: z.number().int(),
      })
    )
    .mutation(async ({ ctx, input }) => {
      // Optimistic concurrency: bump version only if the row's version matches.
      const encryptedDoc = input.document
        ? (encryptDocumentSecrets(input.document) as never)
        : undefined;
      try {
        const updated = await prisma.workflow.updateMany({
          where: { id: input.id, ownerId: ctx.user.id, version: input.version },
          data: {
            ...(input.name ? { name: input.name } : {}),
            ...(encryptedDoc ? { document: encryptedDoc } : {}),
            version: { increment: 1 },
          },
        });
        if (updated.count === 0) {
          // Either not found/forbidden, or version mismatch (stale write).
          const existing = await prisma.workflow.findFirst({
            where: { id: input.id, ownerId: ctx.user.id },
            select: { version: true },
          });
          if (!existing) throw new TRPCError({ code: 'NOT_FOUND' });
          throw new TRPCError({
            code: 'CONFLICT',
            message: `Stale version: current is ${existing.version}, you sent ${input.version}`,
          });
        }
      } catch (e) {
        if (e instanceof TRPCError) throw e;
        throw new TRPCError({ code: 'INTERNAL_SERVER_ERROR', cause: e });
      }
      const wf = await prisma.workflow.findFirst({
        where: { id: input.id },
        select: { id: true, version: true, updatedAt: true },
      });
      return { id: wf!.id, version: wf!.version, updatedAt: wf!.updatedAt };
    }),

  delete: protectedProcedure
    .input(z.object({ id: z.string() }))
    .mutation(async ({ ctx, input }) => {
      const deleted = await prisma.workflow.deleteMany({
        where: { id: input.id, ownerId: ctx.user.id },
      });
      if (deleted.count === 0) throw new TRPCError({ code: 'NOT_FOUND' });
      return { ok: true };
    }),
});

File: apps/flow-backend-python/app/api/routers/workflow.py
"""Workflow CRUD router — mirrors Node routers/workflow.ts.

Routes are RESTful under /api/workflow/* (NOT tRPC). Phase 7 will switch the
editor's api/trpc.ts client to call these paths. Business logic (optimistic
concurrency, secret encryption/decryption) matches Node one-for-one.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.core.auth import CurrentUser, require_user
from app.core.database import get_db
from app.models import Workflow
from app.utils.secrets import decrypt_document_secrets, encrypt_document_secrets

from ..schemas import (
    IdIn,
    WorkflowCreateIn,
    WorkflowCreateOut,
    WorkflowDeleteOut,
    WorkflowFull,
    WorkflowListItem,
    WorkflowUpdateIn,
    WorkflowUpdateOut,
)

router = APIRouter(prefix="/api/workflow", tags=["workflow"])


@router.get("/list", response_model=list[WorkflowListItem], response_model_by_alias=True, dependencies=[Depends(require_user)])
def list_workflows(
    user: CurrentUser = Depends(require_user),
    search: str | None = Query(default=None),
    db: Session = Depends(get_db),
) -> list[Workflow]:
    """List the caller's workflows, newest first. Omits document body."""
    stmt = select(Workflow).where(Workflow.owner_id == user.id)
    if search:
        stmt = stmt.where(Workflow.name.contains(search))
    stmt = stmt.order_by(Workflow.updated_at.desc())
    return list(db.scalars(stmt))


@router.get("/get", response_model=WorkflowFull, response_model_by_alias=True, dependencies=[Depends(require_user)])
def get_workflow(
    id: str = Query(...),
    user: CurrentUser = Depends(require_user),
    db: Session = Depends(get_db),
) -> Workflow:
    """Fetch one workflow (with document, secrets decrypted)."""
    stmt = select(Workflow).where(Workflow.id == id, Workflow.owner_id == user.id)
    wf = db.scalars(stmt).first()
    if wf is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Workflow not found")
    # Decrypt secrets in a copy so the cached ORM object stays encrypted.
    wf.document = decrypt_document_secrets(wf.document)
    return wf


@router.post("/create", response_model=WorkflowCreateOut, dependencies=[Depends(require_user)])
def create_workflow(
    payload: WorkflowCreateIn,
    user: CurrentUser = Depends(require_user),
    db: Session = Depends(get_db),
) -> WorkflowCreateOut:
    """Create a workflow. Secrets in document are encrypted before persistence."""
    wf = Workflow(
        name=payload.name,
        owner_id=user.id,
        document=encrypt_document_secrets(payload.document),
        version=1,
    )
    db.add(wf)
    try:
        db.commit()
    except IntegrityError as e:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e.orig)) from e
    db.refresh(wf)
    return WorkflowCreateOut(id=wf.id, version=wf.version)


@router.post("/update", response_model=WorkflowUpdateOut, response_model_by_alias=True, dependencies=[Depends(require_user)])
def update_workflow(
    payload: WorkflowUpdateIn,
    user: CurrentUser = Depends(require_user),
    db: Session = Depends(get_db),
) -> WorkflowUpdateOut:
    """Update a workflow with optimistic concurrency (version must match).

    Mirrors Node's updateMany-based CAS: bump version only if the row's current
    version matches the client-sent version. Stale writes → 409 CONFLICT.
    """
    stmt = select(Workflow).where(Workflow.id == payload.id, Workflow.owner_id == user.id)
    wf = db.scalars(stmt).first()
    if wf is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Workflow not found")
    if wf.version != payload.version:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=f"Stale version: current is {wf.version}, you sent {payload.version}",
        )

    if payload.name is not None:
        wf.name = payload.name
    if payload.document is not None:
        wf.document = encrypt_document_secrets(payload.document)
    wf.version += 1
    db.commit()
    db.refresh(wf)
    return WorkflowUpdateOut(id=wf.id, version=wf.version, updated_at=wf.updated_at)


@router.post("/delete", response_model=WorkflowDeleteOut, dependencies=[Depends(require_user)])
def delete_workflow(
    payload: IdIn,
    user: CurrentUser = Depends(require_user),
    db: Session = Depends(get_db),
) -> WorkflowDeleteOut:
    """Delete a workflow. Idempotent on auth-scoped rows."""
    stmt = select(Workflow).where(Workflow.id == payload.id, Workflow.owner_id == user.id)
    wf = db.scalars(stmt).first()
    if wf is None:
        # Match Node: NOT_FOUND (row absent or not owned by caller).
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Workflow not found")
    db.delete(wf)
    db.commit()
    return WorkflowDeleteOut(ok=True)
